### PR TITLE
Feature/plugin repo

### PR DESCRIPTION
**Pull request by novaknole**
## Description

Please include a summary of the change and be sure you follow the contributions rules we do provide [here](./CONTRIBUTIONS.md)

Task: [ID]()

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: packages/contracts/contracts/plugin/IPluginRepo.sol [hunk 1]
```diff
@@ -6,12 +6,8 @@ pragma solidity 0.8.10;
 /// @author Aragon Association - 2022
 /// @notice The interface required for a plugin repository.
 interface IPluginRepo {
-    /// @notice Creates a new version with contract `_pluginSetupAddress` and content `@fromHex(_contentURI)`.
-    /// @param _semanticVersion The semantic version number for the version to be created.
-    /// @param _pluginSetupAddress The address of the plugin setup contract.
-    /// @param _contentURI External URI where the plugin metadata and subsequent resources can be fetched from
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/IPluginRepo.sol:12]: Why removing these comments?

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 2]
```diff
@@ -27,10 +27,12 @@ contract PluginRepo is
...
 {
     using Address for address;
-
+    
     struct Version {
-        uint16[3] semanticVersion;
+        uint8 releaseId;
+        uint16 buildId;
         address pluginSetup;
+        bool verified;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginRepo.sol:35]: Verified like... by the community or someone? This feels out of scope by now

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 3]
```diff
@@ -40,21 +42,23 @@ contract PluginRepo is
...
-    mapping(uint256 => Version) internal versions;
+    /// @notice increasing build ids per release
+    /// @dev keys can only be uint8 even though uint256 is specified to reduce gas cost.
+    mapping(uint256 => uint256) internal buildIdsPerRelease;
 
-    /// @notice A mapping between the semantic version number hash and the version index.
-    mapping(bytes32 => uint256) internal versionIndexForSemantic;
+    /// @notice The mapping between version hash and its corresponding Version information.
+    /// @dev key: keccak(abi.encode(releaseId, buildId)) value: Version struct.
+    mapping(bytes32 => Version) internal versions;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginRepo.sol:54]: do we need a version hash here? wouldn't it be easier to have a 2-level awway that would be iterated like `versions[release][build]`?

Maybe it's more efficient to just hash releaseId+buildId and index using subgraph :+1:

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 4]
```diff
@@ -118,110 +168,106 @@ contract PluginRepo is
             revert InvalidPluginSetupInterface({invalidPluginSetup: _pluginSetup});
         }
 
-        uint256 currentVersionIndex = nextVersionIndex - 1;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginRepo.sol:121]: Does it bring value to spend gas in this condition? :thinking: 
- Maybe a developer wants to do a prerelease (but this should happen on testnet, not in prod)
- But maybe we need to ensure that nobody created a build on an empty release

File: packages/contracts/contracts/plugin/PluginRepo.sol [hunk 5]
```diff
@@ -118,110 +168,106 @@ contract PluginRepo is
...
-    /// @notice Gets the version information associated with a semantic version number.
-    /// @return semanticVersion The semantic version number.
-    /// @return pluginSetup The address of the plugin factory associated with the version.
-    /// @return contentURI The external URI pointing to the content of the version.
-    function getVersionBySemanticVersion(uint16[3] memory _semanticVersion)
+    /// @notice get the version by release id and build id.
+    /// @param _releaseId the release id
+    /// @param _buildId the build id in _releaseId
+    /// @return Version the version which is binded to the hash of (_releaseId, _buildId)
+    function versionByReleaseAndBuildId(uint8 _releaseId, uint16 _buildId)
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginRepo.sol:233]: How about having less functions, doing a simpler job?
- `getVersionHash(release, build)`
- `getVersion(versionHash)`

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 6]
```diff
@@ -211,7 +211,7 @@ contract PluginSetupProcessor is DaoAuthorizable {
         }
 
         // Reverts if pluginSetup doesn't exist on the repo...
-        _pluginSetupRepo.getVersionByPluginSetup(_pluginSetup);
+        _pluginSetupRepo.latestVersionByPluginSetup(_pluginSetup);
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:214]: how about `getLastPluginSetupVersion()`?

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 7]
```diff
@@ -339,19 +339,18 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
                 .pluginSetupRepo
-                .getVersionByPluginSetup(_updateParams.newPluginSetup);
+                .latestVersionByPluginSetup(_updateParams.newPluginSetup);
 
-            // Assert that the version bump valid
-            if (!isValidBumpLoose(oldVersion, newVersion)) {
-                revert BumpInvalid({currentVersion: oldVersion, nextVersion: newVersion});
+            if(currentVersion.releaseId != newVersion.releaseId) {
+                // revert not allowed
             }
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:354]: Should a nop be kept?

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 8]
```diff
@@ -364,7 +363,7 @@ contract PluginSetupProcessor is DaoAuthorizable {
                 _dao,
                 _updateParams.plugin,
                 _currentHelpers,
-                oldVersion,
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:367]: Don't we need the release? Or we assume that the plugin setup will always be on the same release, I guess

File: packages/contracts/contracts/plugin/examples/count/CounterV2PluginSetup.sol [hunk 9]
```diff
@@ -125,8 +125,9 @@ contract CounterV2PluginSetup is PluginSetup {
         )
     {
         uint256 _newVariable;
-
-        if (_oldVersion[0] == 1 && _oldVersion[1] == 0) {
+        
+        // if (_oldVersion[0] == 1 && _oldVersion[1] == 0) { // TODO:GIORGI
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/examples/count/CounterV2PluginSetup.sol:129]:
```suggestion
```

File: packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol [hunk 10]
```diff
@@ -114,7 +114,8 @@ contract PluginUUPSUpgradeableSetupV2Mock is PluginUUPSUpgradeableSetupV1Mock {
         )
     {
         // Update from V1
-        if (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) {
+        if(_currentBuildId == 1) {
+        // if (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) { // TODO:GIORGI
```
Comment (brickpop) [inline packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:118]:
```suggestion
```

File: packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol [hunk 11]
```diff
@@ -165,7 +166,8 @@ contract PluginUUPSUpgradeableSetupV3Mock is PluginUUPSUpgradeableSetupV2Mock {
         (_dao, _plugin, _helpers);
 
         // Update from V1
-        if (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) {
+        if(_currentBuildId == 1) {
+        // if (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) { // TODO:GIORGI
```
Comment (brickpop) [inline packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:170]:
```suggestion
```

File: packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol [hunk 12]
```diff
@@ -175,7 +177,8 @@ contract PluginUUPSUpgradeableSetupV3Mock is PluginUUPSUpgradeableSetupV2Mock {
         }
 
         // Update from V2
-        if (_oldVersion[0] == 1 && _oldVersion[1] == 1 && _oldVersion[2] == 0) {
+        if(_currentBuildId == 2) {
+        // if (_oldVersion[0] == 1 && _oldVersion[1] == 1 && _oldVersion[2] == 0) { // TODO:GIORGI
```
Comment (brickpop) [inline packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:181]:
```suggestion
```

File: packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol [hunk 13]
```diff
@@ -165,7 +166,8 @@ contract PluginUUPSUpgradeableSetupV3Mock is PluginUUPSUpgradeableSetupV2Mock {
         (_dao, _plugin, _helpers);
 
         // Update from V1
-        if (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) {
+        if(_currentBuildId == 1) {
```
Comment (brickpop) [inline packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:169]: I would use examples in terms of ranges (`_currentBuildId <= 14`)

File: packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol [hunk 14]
```diff
@@ -236,22 +239,25 @@ contract PluginUUPSUpgradeableSetupV4Mock is PluginUUPSUpgradeableSetupV3Mock {
...
         // (1.1.0 or 1.0.0), that means logic contracts change and It's required 
         // to call initialize for the upgrade call. Logic below is just a test 
         // but dev is free to do what he wishes.
         else if (
-            (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) ||
-            (_oldVersion[0] == 1 && _oldVersion[1] == 1 && _oldVersion[2] == 0)
+            // (_oldVersion[0] == 1 && _oldVersion[1] == 0 && _oldVersion[2] == 0) ||
+            // (_oldVersion[0] == 1 && _oldVersion[1] == 1 && _oldVersion[2] == 0)
+            _currentBuildId == 1 ||
+            _currentBuildId == 2
```
Comment (brickpop) [inline packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:254]: `else if (_currentBuildId < 3)`

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 15]
```diff
@@ -32,11 +32,56 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
+    }
+
+    /// @notice the struct containing params for the apply installation function.
+    struct PrepareUpdateParams {
         address plugin; // The address of the `Plugin` contract being currently active. This can be a proxy or a concrete instance.
         PluginRepo pluginSetupRepo; // The repository storing the `PluginSetup` contracts of all versions of a plugin.
-        address currentPluginSetup; // The `PluginSetup` contract of the version being currently active and to be updated from.
-        address newPluginSetup; // The `PluginSetup` contract of the version to be updated to.
+        PluginRepo.Tag currentVersionTag; // The `current` version of the plugin being currently active and to be updated from.
+        PluginRepo.Tag newVersionTag; // The `new` version of the plugin to be updated to.
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:55]: Does this allow us to prevent jumping from release1-build-5 to release1-build-2?

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 16]
```diff
@@ -32,11 +32,56 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
+        address[] currentHelpers;
+        bytes data;
+    }
+
+    /// @notice the struct containing params for the apply installation function.
+    struct ApplyUpdateParams {
+        address plugin;
+        PluginRepo.Tag versionTag;
+        PluginRepo pluginSetupRepo;
+        bytes initData;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:65]: Is this `initData` fundamentally different from the `bytes data` above?

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 17]
```diff
@@ -32,11 +32,56 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
+        PermissionLib.ItemMultiTarget[] permissions;
+    }
+
+    /// @notice the struct containing params for the apply installation function.
+    struct PrepareUninstallParams {
+        address plugin;
+        PluginRepo.Tag versionTag;
+        PluginRepo pluginSetupRepo;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:73]:
```suggestion
        PluginRepo pluginRepo;
```

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 18]
```diff
@@ -32,11 +32,56 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
+        PluginRepo.Tag newVersionTag; // The `new` version of the plugin to be updated to.
+        address[] currentHelpers;
+        bytes data;
+    }
+
+    /// @notice the struct containing params for the apply installation function.
+    struct ApplyUpdateParams {
+        address plugin;
+        PluginRepo.Tag versionTag;
+        PluginRepo pluginSetupRepo;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:64]:
```suggestion
        PluginRepo pluginRepo;
```

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 19]
```diff
@@ -32,11 +32,56 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
+        address[] currentHelpers;
+        bytes data;
+    }
+
+    /// @notice the struct containing params for the apply uninstall function.
+    struct ApplyUninstallParams {
+        address plugin;
+        PluginRepo.Tag versionTag;
+        PluginRepo pluginSetupRepo;
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:82]:
```suggestion
        PluginRepo pluginRepo;
```

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 20]
```diff
@@ -101,15 +146,17 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
+    /// @param versionTag The version tag of the plugin to used for install preparation.
     /// @param data The `bytes` encoded data containing the input parameters for the installation as specified in the `prepareInstallationDataABI()` function in the `pluginSetup` setup contract.
     /// @param plugin The address of the plugin contract.
     /// @param helpers The address array of all helpers (contracts or EOAs) that were prepared for the plugin to be installed.
     /// @param permissions The list of multi-targeted permission operations to be applied to the installing DAO.
     event InstallationPrepared(
         address indexed sender,
         address indexed dao,
-        address indexed pluginSetup,
+        PluginRepo indexed pluginSetupRepo,
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:158]: Same as above here

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 21]
```diff
@@ -231,34 +299,31 @@ contract PluginSetupProcessor is DaoAuthorizable {
         emit InstallationPrepared({
             sender: msg.sender,
             dao: _dao,
-            pluginSetup: _pluginSetup,
-            data: _data,
+            pluginSetupRepo: params.pluginSetupRepo,
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:302]:
```suggestion
            pluginRepo: params.pluginRepo,
```

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 22]
```diff
@@ -267,91 +332,95 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
-     * and checking it always would cost more... shall we still check it and how ? */
-    /// @notice Prepares the update of an UUPS upgradeable plugin.
-    /// @param _dao The address of the installing DAO.
-    /// @param _updateParams The parameters of the update.
-    /// @param _currentHelpers The address array of all current helpers (contracts or EOAs) associated with the plugin that is prepared to be updated.
-    /// @param _data The `bytes` encoded data containing the input parameters for the update as specified in the `prepareUpdateDataABI()` function in the `newPluginSetup` setup contract inside `_updateParams`.
-    /// @return permissions The list of multi-targeted permission operations to be applied to the updating DAO.
-    /// @return initData The initialization data to be passed to upgradeable contracts when the update is applied
-    /// @dev The list of `_currentHelpers` has to be specified in the same order as they were returned from previous setups preparation steps (the latest `prepareInstallation` or `prepareUpdate` step that has happend) on which the update is prepared for.
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:297]: Were these comments moved to the newly created structs?

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 23]
```diff
@@ -267,91 +332,95 @@ contract PluginSetupProcessor is DaoAuthorizable {
...
-        if (!repoRegistry.entries(address(_updateParams.pluginSetupRepo))) {
+        if (!repoRegistry.entries(address(params.pluginSetupRepo))) {
             revert PluginRepoNonexistent();
         }
 
         // Check if plugin is applied
-        if (!isInstallationApplied[_getAppliedId(_dao, _updateParams.plugin)]) {
+        if (!isInstallationApplied[_getAppliedId(_dao, params.plugin)]) {
             revert SetupNotApplied();
         }
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:385]: I'm not sure I follow the logic here (though it's not new). Preparing an update needs an installation to be applied. Is this relevant at this point?

File: packages/contracts/contracts/plugin/PluginSetupProcessor.sol [hunk 24]
```diff
@@ -386,9 +455,10 @@ contract PluginSetupProcessor is DaoAuthorizable {
         emit UpdatePrepared({
             sender: msg.sender,
             dao: _dao,
-            pluginSetup: _updateParams.newPluginSetup,
-            data: _data,
-            plugin: _updateParams.plugin,
+            pluginSetupRepo: params.pluginSetupRepo,
```
Comment (brickpop) [inline packages/contracts/contracts/plugin/PluginSetupProcessor.sol:458]:
```suggestion
            pluginRepo: params.pluginRepo,
```

**Comment by Michael-A-Heuer** [comment]
We should consider adding more structure to give more clarity.

- make additional, nested structs for the arguments returned by the `prepareXYZ` method that go into `applyXYZ` methods. This would make it less confusing where all the arguments come from

- make a struct for those parameters that are present in all 6 methods and that are used for the `setupId` hash
```
    /// @param _dao The address of the installing DAO.
    /// @param _pluginSetup The address of the `PluginSetup` contract.
    /// @param _pluginSetupRepo The repository storing the `PluginSetup` contracts of all versions of a plugin.
```
`_getSetupId(_dao, _pluginSetup, address(_pluginSetupRepo), _plugin);`
Note that `_plugin` is not present in all 6 methods, because it doesn't exist before the installation
